doc/src/core-transmutation.md

```diff
@@ -0,0 +1,143 @@
+Challenge: Preconditions for core transmuting methods
+
+## Status
+
+\[Open | Resolved | Expired\]: Open
+
+## Dates
+
+Start date:
+
+End date:
+
+## Goal
+
+Confirm the correctness of value transmutations performed by libcore, including those transmutations exposed as public methods provided by libcore.
+
+To keep the goal somewhat manageable, it excludes some classes of code (e.g. utf8-validation, async tasks, and others); see the assumptions for the full list of excluded categories.
+
+## Details
+
+### Motivating Example
+
+There are several calls to these unsafe methods within the code of safe methods exported libcore itself, so having clear and verifiable safety contracts is critical for verifying the safety of the safe methods that invoke them.
+
+For example, `std::slice::align_to` (which unsafely converts any slice `&[T]` into a tuple `(&[T], &[U], &[T])` composed of a prefix, a maximal sequence of values of type `U`, and a suffix) just says for its safety that “all the usual caveats pertaining to `transmute::<T, U>` also apply here”, but the documentation for `transmute` is frankly (and rightfully) scary.
+
+### Part 1: the two intrinsics
+
+The public unsafe intrinsic `transmute<T, U>` takes a value of type `T` and reinterprets it to have type `U`, with the sole (statically-enforced) restriction being that the types `T` and `U` must have the same size. The unstable intrinsic `transmute_unchecked` is similar, except that it removes the static size restriction, treating violations of it as undefined behavior.
+
+What is the right way to encode the preconditions of the two transmutation intrinsics?
+
+If one is solely concerned about safety: The Rustonomicon lists several ways that transmutation can yield undefined behavior. Encoding a correct safety contract for transmute is crucial.
+
+If one is concerned about proving functional correctness, then reasoning about transmutation will require reasoning about the byte representation of values to justify that the reinterpretation of the value’s bytes for satisfying the pending proof obligation associated with the output value.
+
+
+### Part 2: unsafe APIs with validity constraints
+
+There are unsafe methods (which are defined by libcore and reexported by libstd) have the effect of a transmutation between a (sequence of) T and a (sequence of) U. Come up with an appropriate safety contract for each of them; they should usually be something less arduous for callers than transmute itself (we hope).
+
+
+### Part 3: unsafe APIs with richer constraints
```

Comment:
Do the  "parts" correspond to different subsets of the functions found in core ?

Reply:
I assume @pnkfelix split the set of all functions that should be verification targets into the core `transmute` intrinsics, the subset of unsafe functions that rely on `transmute` that only contains validity constraints, the subset of unsafe functions with functional constraints, and the subset of safe functions that rely on transmute.

Reply:
Hey that actually does sound like what I was thinking, but phrased in a really clean way!